RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,615 @@
+      value of the hybrid tensor, hence, the memory footprint of a
+      sparse tensor will be slightly smaller,
+    - evaluation of a element-wise function on sparse tensors will be
+      optimal, see point 8 below,
+    - and most importantly, the detection of zero fill value of a
+      hybrid tensor will be optimal, see point 10 below.
+
+7.  The fill value of a sparse tensor can be changed in place.
+
+    For instance:
+
+    ```python
+    A.fill_value().fill_(1.2)
+    ```
+
+    resets the fill value of a sparse tensor `A` to be `1.2`. Note
+    that one can use both `fill_value()` or `_fill_value()` outputs
+    for resetting the fill value of the given tensor.
+
+8.  If `A` is a (coalesced) sparse tensor and `f` is any calculus
+    function that is applied to a tensor element-wise, then:
+
+    ```python
+    f(A) == torch.sparse_coo_tensor(A.indices(), f(A.values()), fill_value=f(A._fill_value()))
+    ```
+
+    From this relation follows an identity:
+
+    ```python
+    f(A).to_dense() == f(A.to_dense())
+    ```
+
+    that will be advantageous in testing the sparse tensor support of
+    element-wise functions.
+
+9.  The fill value of an element-wise n-ary operation on sparse
+    tensors with different fill values is equal to the result of
+    applying the operation to the fill values of the sparse
+    tensors.
+
+    For instance (`*` represents unspecified element),
+
+    ```
+    A = [[1, *],
+         [3, *]]
+    B = [[5, *],
+         [*, 8]]
+    # assume A and B fill values are 2 and 6, respectively, then
+    A + B = [[6,  *],
+             [9, 10]]
+    # with fill value 2 + 6 = 8
+    ```
+
+10. Existing PyTorch functions that support sparse tensor as inputs,
+    need to be updated for handling the defined fill values. This will
+    be implemented in two stages:
+
+    1. All relevant functions need to check for zero fill value. If a
+       nonzero fill value is used, raise an exception.
+
+       For instance:
+
+       ```C++
+       TORCH_CHECK(A._fill_value().nonzero().numel() == 0,
+                   "The <function> requires a sparse tensor with zero fill value, got ",
+                   A.fill_value());
+       ```
+
+       This check ensures that the existing functions will not produce
+       incorrect results silently when users pass a sparse tensor
+       with nonzero fill value as an input while the function assumes
+       it to be zero.
+
+    2. Update the related functions to handle nonzero fill values of
+       sparse tensor inputs correctly.
+
+       For instance, consider a matrix multiplication of two sparse
+       tensors `A` and `B` with fill values `a` and `b`, respectively,
+       then the `matmul` operation can be expanded as follows:
+
+       ```python
+       matmul(A, B) = matmul(A - a + a, B - b + b)
+                    = matmul(A - a, B - b) + a * matmul(ones_like(A), B) + b * matmul(A, ones_like(B))
```

Comment:
cc @albanD doesn't this look like dual numbers? It kind of looks like dual numbers.

Reply:
Does this actually preserve sparsity? Consider this example:

```
>>> torch.tensor([[1,0,0],[0,2,0],[0,0,3]], dtype=torch.float) @ torch.ones((3, 3))
tensor([[1., 1., 1.],
        [2., 2., 2.],
        [3., 3., 3.]])
```

I'm not sure the output can be expressed in a sparse way. (I suppose the transpose can be done as one sparse dimension and one dense dimension, with a fill value of `[1, 2, 3]`)

Reply:
No, in general, the matmul of sparse tensors with nonzero fill values will result in a dense tensor (read: a tensor in sparse format but with a small number of equal elements).

While this is going to be slightly off-topic, but when there exist applications that require such a matmul, I can think of one way to preserve the sparsity from matmul: a lazy evaluation. It would be like uncoalesced tensor but at the level of tensors, not tensor elements.

For instance, the result of `A @ B` would be an object, say, LazyTensor that holds three sparse tensors: `(A-a) @ (B-b)`, `ones_like(A) @ (a*B)`, and `(b*A) @ ones_like(B)`. Each of these tensors can be stored memory-efficiently as sparse tensors while adding these would lead to a dense tensor (but we are going to postpone the addition). So, when one has
```
A @ B -> C = LazyTensor((A-a) @ (B-b), ones_like(A) @ (a*B), (b*A) @ ones_like(B))
```
then computing a matrix-vector product `C @ v` boils down to computing the sparse matrix-vector product with each of the three sparse tensors and `v`:
```
LazyTensor(C1, C2, C3) @ v -> C1 @ v + C2 @ v + C3 @ v
```
so that the postponed addition is realized with the results of matrix-vector products.

Reply:
OK. We're generally amenable to laziness (see also https://github.com/pytorch/pytorch/issues/43563 ) and it is easier to do this in sparse as we have no concept of views, which make laziness harder to do. But out of scope for this RFC :)